### PR TITLE
Issue: #7733 Update AbstractChecks to log DetailAST - ImportOrder

### PR DESCRIPTION
**Pull request by DXTkastb**
Fixes issue #7733 

**Updates(total 5 log calls):**
https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L666

https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L673

https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L676

https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L738

https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L754

**diff report:**
https://dxtkastb.github.io/checkstyle/reports/ImportOrder/diff/

### CODE REVIEW

**Comment by DXTkastb** [comment]
@rnveach Please review.

**Comment by DXTkastb** [comment]
Sorry, I accidentally pushed unwanted changes to this PR, I will correct this soon.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/DefaultComesLastCheck.java [hunk 1]
```diff
@@ -48,6 +48,48 @@
  * &lt;module name=&quot;DefaultComesLast&quot;/&gt;
  * </pre>
  * <p>
+ * Example:
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/DefaultComesLastCheck.java]: Why are these changes here? They are not related to ImportOrder check.

File: src/xdocs/config_coding.xml [hunk 2]
```diff
@@ -749,6 +749,48 @@ class K {
         </p>
         <source>
 &lt;module name=&quot;DefaultComesLast&quot;/&gt;
+        </source>
```
Comment (rdiachenko) [inline src/xdocs/config_coding.xml]: Not related to ImportOrder check.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/DefaultComesLastCheck.java (hunk 1 above)
Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/DefaultComesLastCheck.java]: @rdiachenko I will revert the changes, I accidentally pushed changes for different issue to this PR

**Comment by DXTkastb** [comment]
@rdiachenko I have now removed the changes, sorry for that. Please review the PR.

File: src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderOne.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package org.checkstyle.suppressionxpathfilter.importorder;
+
+import static java.lang.Math.PI;
+import java.util.Set; //warn
+
+public class SuppressionXpathRegressionImportOrderOne { 
+	// code
+ }
```
Comment (rdiachenko) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderOne.java]: remove trailing whitespace

File: src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderOne.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package org.checkstyle.suppressionxpathfilter.importorder;
+
+import static java.lang.Math.PI;
+import java.util.Set; //warn
```
Comment (rdiachenko) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderOne.java]: add a whitespace between `//` and `warn`, same for all other cases

File: src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderTwo.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package org.checkstyle.suppressionxpathfilter.importorder;
+
+import java.util.List;
+
+import java.util.Set; //warn
+
+public class SuppressionXpathRegressionImportOrderTwo { 
+	// code
+ }
```
Comment (rdiachenko) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderTwo.java]: remove trailing whitespace

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 6]
```diff
@@ -655,25 +654,26 @@ else if (option == ImportOrderOption.INFLOW) {
...
-    private void doVisitToken(FullIdent ident, boolean isStatic, boolean previous, int line) {
+    private void doVisitToken(FullIdent ident, boolean isStatic, boolean previous, DetailAST ast) {
         final String name = ident.getText();
         final int groupIdx = getGroupNumber(isStatic && staticImportsApart, name);
 
         if (groupIdx > lastGroup) {
-            if (!beforeFirstImport && line - lastImportLine < 2 && needSeparator(isStatic)) {
-                log(line, MSG_SEPARATION, name);
+            if (!beforeFirstImport && ast.getLineNo()
+                - lastImportLine < 2 && needSeparator(isStatic)) {
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: could you format it in this way:
```
if (!beforeFirstImport
    && ast.getLineNo() - lastImportLine < 2
    && needSeparator(isStatic))
```

**Comment by DXTkastb** [comment]
done.

**Comment by rdiachenko** [comment]
@DXTkastb could you reply 'done' to each review note please. Otherwise, it is not clear what was fixed.

**Comment by DXTkastb** [comment]
> @DXTkastb could you reply 'done' to each review note please. Otherwise, it is not clear what was fixed.

Sorry @rdiachenko, I'll keep this in mind.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 6 above)
Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: done.

File: src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderTwo.java (hunk 5 above)
Comment (DXTkastb) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderTwo.java]: done

File: src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderOne.java (hunk 4 above)
Comment (DXTkastb) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderOne.java]: done

File: src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderOne.java (hunk 3 above)
Comment (DXTkastb) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderOne.java]: done

File: src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionImportOrderTest.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        final List<String> expectedXpathQueries = Collections.singletonList(
+                "/IMPORT[./DOT/IDENT[@text='Set']]"
+        );
+
+        runVerifications(moduleConfig, fileToProcess, expectedViolation,
+                expectedXpathQueries);
+    }
+
+    @Test
+    public void testThree() throws Exception {
```
Comment (rnveach) [inline src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionImportOrderTest.java]: > Updates(total 5 log calls):

We should have 5 xpath tests, 1 for each log call. Current test ends at 3. It is missing 2 more. Make sure each xpath test is specific to each log.

Comment (DXTkastb) [reply src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionImportOrderTest.java]: Ok, @rnveach I'll add more tests and produce the required regression reports. Thanks for the review

**Comment by rnveach** [comment]
Make sure you keep the number of commits at 1.

**Comment by DXTkastb** [comment]
Please ignore I merged other commits.

**Comment by rnveach** [comment]
Please dont close and recreate the PR. It causes more confusion and we track of what is being requested.
